### ISSUE TITLE
support Win8 / win10 / modern / metro / uwp apps

### ISSUE THREAD

**Issue by jesperbjerke** [Wox-launcher/Wox#198]

**Comment by orzFly** [comment]
https://github.com/luisrigoni/metro-apps-list
http://stackoverflow.com/questions/18219915/get-localized-friendly-names-for-all-winrt-metro-apps-installed-from-wpf-applica

FYI

**Comment by qianlifeng** [comment]
I don't have Win8/10 environment now, I will try above solutions when I got one.

**Comment by zizhengwu** [comment]
I am wondering whether this feature is under development. If implemented, it would be a huge advantage over Hain (https://github.com/appetizermonster/hain) for example.

**Comment by bao-qian** [comment]
WIP:

# notes:

https://github.com/ljw1004/uwp-desktop
https://blogs.msdn.microsoft.com/lucian/2015/10/23/how-to-call-uwp-apis-from-a-desktop-vbc-app/

http://stackoverflow.com/a/21610315/2833083
http://nicksnettravels.builttoroam.com/post/2015/07/24/launching-windows-store-applications-from-code-in-windows-10.aspx

http://stackoverflow.com/questions/32122679/getting-icon-of-modern-windows-app-from-a-desktop-application

# AppxPackaging

http://clrprofiler.codeplex.com/
https://github.com/jimschubert/clr-profiler/blob/master/src/CLRProfiler45Source/WindowsStoreAppHelper/WindowsStoreAppHelper.cs
CLRProfiler45Source\MidlOut\readme.txt

```
cd dir
midl "C:\Program Files (x86)\Windows Kits\8.1\Include\winrt\AppxPackaging.idl"
tlbimp AppxPackaging.tlb /namespace:AppxPackaging /out:AppxPackagingTlb.dll
```

# what have done
1. Name
   1.1 `AppxManifest.xml` -> `Properties` -> `DisplayName`
   1.2 `AppxManifest.xml` -> `xxx:VisualElements` -> `DisplayName`
   1.3 `package.Id.Name`
2. `ms-resource` in 1.1 and 1.2:
   2.1 `SystemApps` -> `pris` -> `resources.lan.pri` -> `ms-resource:xxx`
   2.2 `WindowsApps` -> `resources.pri` -> `ms-resource:xxx`
   2.3 `WindowsApps` -> all `xxx.language-en_xxx` in `WindowsApps` directory -> `resources.pri` -> `ms-resource:xxx`
3. using `FindPackagesForUser` to get all packages
4. parse `AppxManifest.xml` in 1
5. parse different `*.pri` files in 2 using `makepri.exe`. `IndexFileContentsAsync` can't be used because of problems 2 and 3

# problems
1. if we want to use `await` with `IAsyncOperation`, then we must reference `System.Runtime.Windowsruntime.dll`. and `Windows.winmd`.
   But the size of `windows.winmd` is 4mb, same as wox, too big.
2. `IndexFileContentsAsync` with `Microsoft.Windows.Cortana_cw5n1h2txyewy\pris\resources.en-US.pri` will throw `Value does not fall within the expected range`
3. `C:\Program Files\WindowsApps\Microsoft.BingWeather_4.11.156.0_neutral_split.language-zh-hans_8wekyb3d8bbwe\resources.pri: System.Exception: Load/UnloadPriFiles cannot be used with resource packages. (Exception from HRESULT: 0x80073B26)`
4. `IAppListEntry.LaunchAsync()` will throw `{"The specified module could not be found. (Exception from HRESULT: 0x8007007E)":null}`

# todo
1. will .net 4.6 solve problem 1 and 4?
2. use app id to solve mail calendar mismatch issue

**Comment by talynone** [comment]
I have a plugin that does this that is 99% working, without needing to include the .winmd dependency. I'll try to post it to github tonight.

**Comment by bao-qian** [comment]
@talynone that would be great. I'd like to merge your code.

**Comment by talynone** [comment]
It's still very raw/unorganized as I just wrote it 2 days ago. It mostly works, I want to refine the icon getting logic more and turn some methods into async/parallels tasks, you can check out what I have so far here:

https://github.com/talynone/Wox.Plugin.WindowsUniversalAppLauncher

**Comment by bao-qian** [comment]
I remember `SHLoadIndirectString`  has some problems, I will test again.

**Comment by talynone** [comment]
Well anything can be done alternate ways, but so far it works in resolving the DisplayName properly for all my store applications. I can even remove the dependency on packagemanager by getting the full names via "%LOCALAPPDATA%\Microsoft\Windows\Application Shortcuts" folder and or "HKEY_CURRENT_USER\SOFTWARE\Classes\Local Settings\Software\Microsoft\Windows\CurrentVersion\AppModel\Repository\Packages" registry key.

**Comment by talynone** [comment]
There were a few cases where LoadIndirectString wasn't properly extracting the DisplayName/Description, that has been fixed. So for my computer all of them are resolving properly now. Check repository for latest changes.

**Comment by jhdxr** [comment]
@talynone , could you provide a release version of your plugin so we can help testing it. if you have trouble in packaging, you may only offer the dll file and I'll help to pack it.

**Comment by bao-qian** [comment]
@talynone your code truly inspired me to find a better set of APIs.
I will copy part of your code into the program plugin and add your name in the commit message and release description.

updates to all:
It's finally done, I got a small working program with about 200 lines of code.
UWP support will coming out this weekends, support <del>win8.1+</del> win10.
no manual COM import
no manual DisplayName parsing for multi language.

**Comment by jhdxr** [comment]
@happlebao I'm not sure about the things behind your and @talynone 's code, so I have a question. As i said in #857 , I have an app named `Netease Music` in the startup menu, and if I open it the name displayed in the task manger (process) is `网易云音乐`. so which keyword(s) can I use to find that app? the english name or the chinese name? or both?

**Comment by bao-qian** [comment]
@jhdxr 

<del> should be `网易云音乐`. as same as shown in start menu.   (at least my pc will display  `网易云音乐` instead of `Netease Music` in start menu.).
or `yyy \ wangyiyunyinyue`  for pinyin.  </del>

should be `Netease Music`, same as shown in start menu

**Comment by talynone** [comment]
Did you decide to just catalog "%LOCALAPPDATA%\Microsoft\Windows\Application Shortcuts" folder and parse the .lnk file?

**Comment by bao-qian** [comment]
@jhdxr 
dll from @talynone [UWP.zip](https://github.com/Wox-launcher/Wox/files/423897/UWP.zip)
wox with uwp supprt: https://ci.appveyor.com/project/happlebao/wox/build/1.3.254/artifacts

I know the icon is ugly, but at least it is working...
https://www.tallcomponents.com/blog/opacity-masks-in-wpf

**Comment by talynone** [comment]
Nice! A few things I noticed:

You don't have a few fixes I have implemented: 

Try to search for "calendar" or "mail". In your version "mail" doesn't even show up and "calendar" launches the mail client. In my latest code I enumerate all AppIds that belong to a package with multiple apps defined, then I pass along the AppId at launch time and match against it to make sure it launches the right item in a multi app package.

Also there should be no reason to filter against a publisher. Look at my code that rejects a package if the AppListEntry property is set to "none":

https://github.com/talynone/Wox.Plugin.WindowsUniversalAppLauncher/blob/master/StoreAppLauncher/Helpers/NativeApiManifestHelpers.cs

That way if an app is specified not to show up on the Start Menu, we ignore it and that seems to filter out the nonsense packages. 

You seem to be also grabbing the wrong icon (I grab the right one  95% of the cases, the spec is actually quite complex on multi app packages) : (Look at my GetBestLogo method).

https://msdn.microsoft.com/windows/uwp/controls-and-patterns/tiles-and-notifications-app-assets

Also I have preliminary code to generate an icon cache and overlay the current windows accent color as a background color when none/transparent is defined, or a background color specified by the app manifest when it's defined (Check out console tester app) which is what the start menu does.

**Comment by talynone** [comment]
Also when testing my plugin on a friend's computer it crashed, apparently she had a semi corrupt store app which caused looking up the "InstalledLocation.Path" property to throw an exception. You'll see in my code I have a try catch to just go to the next package if that happens.

**Comment by jhdxr** [comment]
@happlebao thx, @talynone 's plugin works well with 1.3.231 on my laptop. and the question I asked above, `网易云音乐` or `wangyiyunyinyue` does work, but neither `Netease Music` nor `yyy` works. (and I'm curious why `yyy` fails as `wangyiyunyinyue` works fine.

**Comment by bao-qian** [comment]
@jhdxr how about updated wox?

**Comment by bao-qian** [comment]
@talynone 
1. multi application package: yeah, I know this one, there are some code refactoring needs to be done for Program plugin. As you can see in UWP class, the multi application is handled.
2. AppListEntry: that's it, thanks! I found it before, then forgot it, and can't find it until now.
3. icon: you seems [forgot to include ImageHelper](https://github.com/talynone/Wox.Plugin.WindowsUniversalAppLauncher/tree/master/StoreAppLauncher/Helpers) ?
4. InstalledLocation.Path: a package throw exception will be ignored during indexing.

**Comment by talynone** [comment]
Somehow ImageHelper got lost in the shuffle, it's there now.

**Comment by talynone** [comment]
Just noticed this in your code:
var user = WindowsIdentity.GetCurrent()?.User;

That's not needed, if you look at the API docs for FindPackageForUser, passing an empty string is treated as a current user lookup.

So you can instead do this: 

var packages = packageManager.FindPackagesForUser(string.Empty);

**Comment by bao-qian** [comment]
@talynone 
1. icon
check latest build with plated icon: https://ci.appveyor.com/project/happlebao/wox/build/1.3.256/artifacts
2. user
I like write explicit code. in this way, you don't need to check documentation to find the meaning of `string.Empty`  when you read the code again in the future

**Comment by talynone** [comment]
Fair enough, everyone has their styles. A one line comment would take care of any  confusion and prevent unnecessary processing from occurring, especially get current user id which I've seen throw an exception in  a sample app I tried that used the same approach. 

Some things aren't showing up, like Microsoft Edge in the new version.

The icons, when Wox finds it, look pretty nice, might steal some of your code that does that. 

You may want to make the app launch async, When I made the launch async in my plugin it made Wox appear much more performant. Wox shouldn't really need to wait for the app launch to complete before it hides itself.

**Comment by bao-qian** [comment]
1. edge: try restart wox, edge definitely should be there
2. icon: this is a open source code, welcome to steel :)
3. async: probably in the future...

**Comment by bao-qian** [comment]
@jesperbjerke @zizhengwu @jhdxr @nolanlawson @Guaidaodl @uadopo @cH40z-Lord
the uwp support (only win10) is finally here, please try it and confirm it is working:
https://ci.appveyor.com/project/happlebao/wox/build/1.3.271/artifacts

could anybody test this build on win7 or win8? I'd like to know whether it will broke or not.

**Comment by talynone** [comment]
Still doesn't resolve mail/calendar properly. 

Mail launches calendar app and calendar launches mail app.

When I search for "Cut the rope" , it only finds Cut the Rope 2", not "Cut the Rope"

Doesn't find "settings", probably can't find lots more beyond that, I think that's enough to go on.

**Comment by Megasware128** [comment]
Same issues for the mail calendar & calendar mail issue and settings issue.

**Comment by bao-qian** [comment]
@talynone 

> When I search for "Cut the rope" , it only finds Cut the Rope 2", not "Cut the Rope"

you mean you installed two apps, and you can only find one?

**Comment by talynone** [comment]
@happlebao Correct.

**Comment by bao-qian** [comment]
finally some updates from  mail/calendar problems:
- option 1: use  `AppListEntry.LaunchAsync()` 

<del> this api will throw `{"The specified module could not be found. (Exception from HRESULT: 0x8007007E)":null}` when you run .net app with **Any CPU** platform. but works fine when you switch platform to x64. </del>

<del>I guess this is **bug** in windows 10 api. so the support for uwp will limited to windows 10 x64 system.</del>

1. although above bug seems fixed, use option1 can't get background color, so the image can't be platted.
2. if you use `AppDisplayInfo.GetLogo()`, it won't be the same as we want for `Square44x44Logo`

- option2: use displayinfo from `AppListEntry` and parse `GetAppUserModelId` from `IAppxManifestApplication`.

this is the current option. but for multiple application package, the order of apps from `IAppxManifestApplication` and `AppListEntry` is not same and not guaranteed. so you will launch mail when you click calendar.

- option3: only use `IAppxManifestApplication` and manually parsing displayinfo.

this is the best option, it won't needs windows 10 dependency too. But this one is **too hard**.

**Comment by bao-qian** [comment]
do anybody know which package contains multiple application except mail and calendar?

**Comment by bao-qian** [comment]
@talynone @Megasware128 
please try again for mail / calendar problems.
the "cut the rope" problems should also be fixed.

@jhdxr 
could you test mail / calendar uwp apps in chinese local?

https://ci.appveyor.com/project/happlebao/wox/build/1.3.292/artifacts

**Comment by talynone** [comment]
@happlebao mail/calendar works fine now. Thanks. As mentioned before you may want to do the app launch async, so the wox dialog dismisses quickly. 

Btw my code can deal with mail/calendar just fine, and uses no hacks. Look at my app launch code to see how I deal with it.

**Comment by Megasware128** [comment]
Same for me. Calander and Mail fixed

**Comment by quericy** [comment]
@happlebao Hi, I install the `Wox-1.3.292.exe` ,but it doesn't work.
The log file show the wrong folder path of Wox Plugins ,and the memory usage is abnormal:

![memory](https://i.imgur.com/0qjehBm.png)

![log file](https://i.imgur.com/QjkrfUB.png)

**Comment by happlebao** [comment]
> {"The specified module could not be found. (Exception from HRESULT: 0x8007007E)":null}
this bug? seems fixed in latest windows 10(14393.447), at least for me, a new build will be available this weekend.

**Comment by bao-qian** [comment]
hi @quericy 
sorry for late fix,
try again with [1.3.321](https://ci.appveyor.com/project/happlebao/wox/build/1.3.321/artifacts)
upload your whole log file (%APPDATA%\Log\1.3.319\date) once wox startup

**Comment by bao-qian** [comment]
close due to no more new issue.
reopen if you found one.

**Comment by llinfeng** [comment]
I would like to step in with a quick "manual" fix, as I did not find clear instruction on *what to do* from the lengthy discussion above.
* Source: https://lifehacker.com/how-to-create-a-shortcut-for-any-modern-windows-app-1722569853
* Core steps:
    1. To open a magic folder that contains all "icon" to any _registered_ programs on Windows. Note, this thankfully include those troublesome **Windows Store** Apps.
    2. Create shortcuts for these Apps, and give them literal names;
    3. Move those shortcuts to some folder that has been *included* in your `PATH`, and make Wox aware of such folder of shortcuts (the PATH step may be unnecessary, but I will press Win + R at times, when calling those shortcuts that I have created);
![image](https://user-images.githubusercontent.com/5890407/45435778-3c165b00-b67f-11e8-9484-f15b932b66cb.png)
    4. Restarting Wox will do the magic: now the shortcut you've created, however it is named, is several keystrokes away from you.
![image](https://user-images.githubusercontent.com/5890407/45435898-8bf52200-b67f-11e8-87d0-d745dc897906.png)

In case one is curious (as I did myself), here goes what goes into that "shortcut to Windows Store App": something junky, un-editable and un-copy-able.
![image](https://user-images.githubusercontent.com/5890407/45435959-b47d1c00-b67f-11e8-8939-383782df6527.png)

**Comment by Fooughhy** [comment]
Can this become a priority for Wox? I would really like this functionality.

**Comment by bao-qian** [comment]
@Fooughhy 
try new release, it has already been supported.
https://github.com/Wox-launcher/Wox/releases

**Comment by Fooughhy** [comment]
I have a UWP app called "Microsoft Fjärrskrivbord" (Microsoft Remote Desktop as a direct translation) and it is still not showing up though... I just downloaded 1.4.1115.

I reindexed the "Programs" plugin. Do I require anything else?

![Hxss7C6JKp](https://user-images.githubusercontent.com/6977416/82408337-ab0aa080-9a6b-11ea-9cf5-2201063c19d1.jpg)
![YPjUydEkQ9](https://user-images.githubusercontent.com/6977416/82408346-afcf5480-9a6b-11ea-90d2-881da713b179.jpg)
